opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py

```diff
@@ -194,15 +212,33 @@ def merge(self, other: "Resource") -> "Resource":
         """
         merged_attributes = self.attributes
         merged_attributes.update(other.attributes)
-        return Resource(merged_attributes)
+
+        if self.schema_url == "":
+            schema_url = other.schema_url
+        elif other.schema_url == "":
+            schema_url = self.schema_url
+        elif self.schema_url == other.schema_url:
+            schema_url = other.schema_url
+        else:
+            logger.error(
+                "Failed to merge resources: The Schema URL of the old and updating resources are not empty and are different"
+            )
+            return _EMPTY_RESOURCE
```

Comment:
I'm not sure if the empty resource should be returned on merge failed. Maybe we can just leave the `schema_url` as is of the old resource.

Reply:
Since merge fails only when the old and updating schema URLs are different, it's impossible to know whether the merged attributes will conform to the old schema, the updating schema, or neither. There are a few options for dealing with this:
* Return the old resource. The data from the updating resource is lost completely, but this is unlikely to cause cascading errors.
* Return a resource with merged attributes and empty schema URL. The end user loses no data, but this may result in cascading errors where later operations depend on the resource conforming to some schema.
* Return a resource with merged attributes and the old schema URL. The resource doesn't conform to the schema.
* Return an empty resource. This results in total data loss, but since the error is likely caused by the user's incorrect configuration, total data loss would alert them to the problem.

Since you're suggesting the schema URL be preserved from the old resource, the two relevant options are the first and third options. The first causes data loss, but the other will return a resource with invalid state. I think the first option is the best of these 4, but I'd like your opinion on the matter.

Reply:
I also agree that this one is pretty tricky! According to the [specs for merging resources](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#merge), 
> Else this is a merging error (this is the case when the Schema URL of the old and updating resources are not empty and are different).
It doesn't seem to specify what exactly should be returned as the new resource, but I agree that the third option you listed and what @lzchen suggested probably makes the most sense (returning a resource with merged attributes and old schema URL). If you were to return a resource with an empty schema URL, merging a third, different schema URL to this new resource would work without error, which is probably not ideal.

Reply:
https://github.com/tigrannajaryan/opentelemetry-go/blob/084ea75cf5f7cbe888d17109b6bbe1cf4a2e741c/sdk/resource/resource.go#L177

I talked to Tigran to ask about his intention, and [his implementation in Go](https://github.com/tigrannajaryan/opentelemetry-go/blob/084ea75cf5f7cbe888d17109b6bbe1cf4a2e741c/sdk/resource/resource.go#L177) returns an empty resource. He intended for the resulting Resource to have undefined contents to leave it to the implementations to decide what's most natural. This being the case, I think the empty resource is the most natural way to indicate that the Resource is the result of an error.

Reply:
Actually I'd agree with @dgetu 's suggestion of the first option. I'd rather the whole merge do nothing and fail than have the resulting merged resource be a hybrid of the attributes and schema_url (which is possible to have not be consistent). Also, the log.error is sufficient in notifying the user, instead of completely throwing away everything (in the case of the empty resource).

If we are following this way, I think it would also be good in the error message to actually log out the old and new `schema_url`s. This way, it is easier to identify the conflicting urls when doing multiple merges (like when we have a resource detector).

Reply:
Done!